File: insert/generate.py
#!/usr/bin/python2

''' Generates random data for insert.py '''

import random
import time
import json

def generate_entry():
	''' Generates a single entry. Returns a tuple: (id, data)'''

	_id = generate_id()
	flights = random_flights()
	personalflightdata = {}
	passengers = random_passengers(num=random.randint(1, 5))
	fare = random_fare_for(passengers, flights)

	for flight in flights:
		personalflightdata[flight['id']] = random_personal_for(flight)

	return (_id, {
		"id": _id,
		"personalflightdata": personalflightdata,
		"passengers": passengers,
		"flights": flights,
		"fares": fare
	})


# Some generators
letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
digits = '1234567890'

def generate_id(length=6):
	return ''.join(random.sample(letters + digits, length))


def random_flights(count=2):
	return [random_flight() for i in xrange(count)]

def random_personal_for(flight):
	return {
		'class': random.choice(['Economy', 'Business']),
		'baggage': 'ADT %dPC' % random.randint(1, 3)
		# TODO: eTickets
	}

def random_fare_for(passengers, flights):
	name = random.choice(passengers)['id']
	base = len(flights) * 400 + random.randint(0, 100)
	tax = random.uniform(.04, .21)
	taxes = float(base) * tax
	total = round(base + tax, 2)
	curr = random.choice(['EUR', 'USD'])
	return {
			"segment": "*",
			"name": name,
			"type": "ADT",
			"fop": "Invoice",
			"base": base,
			"taxes": taxes, 
			"total": total,
			"currency": curr
	}

names = ['Stegmann', 'Becker', 'Bahro', 'Schreeck', 'Wolf', 'Weber', 'Lovecraft', 'Schaaf']
surnames = ['Patrick', 'Jens', 'Sven', 'Marcel', 'Sandra', 'H.P.', 'Sarah']

def random_passengers(num = 1):
	return [random_passenger() for i in xrange(num)]

def random_passenger():
	name = random.choice(names)
	surname = random.choice(surnames)
	_type = random.randint(0, 2)
	_id = '%s/%s' % (name.upper(), surname.upper())
	if _type == 1:
		_id += 'CHD'
	elif _type == 2:
		_id += 'INF'

	return {
				"id": _id, 
				"type": _type,
				"name": name,
				"surname": surname
	}

# Flight generation - much data to sample
airlines = ['Aeroflot', 'Air France', 'German Wings', 'Air Berlin', 'Lufthansa', 'Aer Lingus',
			'US Airways', 'Delta Airlines']
aircrafts = ['Airbus A380', 'Boeing 747', 'Airbus A320-200']
destinations = [
	{
		"country": "France",
		"city": "Paris",
		"iata":"ORY",
		"airport": "Orly Arpt",
		"terminals": ["W"]
	},
	{
		"country": "France",
		"city": "Montpellier",
		"iata":"MPL",
		"airport": "Frejorgues Arpt",
		"terminals": []
	},
	{
		"country": "Germany",
		"city": "Munich",
		"iata":"MUC",
		"airport": "Munich Intl Arpt",
		"terminal": "2"
	},
	{
			
		"country": "Germany",
		"city": "Dusseldorf",
		"iata":"DUS",
		"airport": "Dusseldorf Arpt",
		"terminal": ""
	},
	{
		"country": "France",
		"city": "Toulouse",
		"iata":"TLS",
		"airport": "Blagnac Arpt",
		"terminal": ""
	},
	{
		"country": "Thailand",
		"city": "Bangkok",
		"iata":"BKK",
		"airport": "Suvarnabhumi Intl Arpt",
		"terminal": ""
	}
]

def random_flight():
	_id = ''.join(random.sample(letters, 2)) + ''.join(random.sample(digits, 4))
	t = int(time.time())
	departure, arrival = (t, random.randrange(t, t+50000))
	duration = arrival-departure
	airline = random.choice(airlines)
	_from = random.choice(destinations)
	_to = random.choice(destinations)

	# Assert that from != to
	while _from == _to:
		_to = random.choice(destinations)

	# In rare cases, add a stop.
	if random.random() < .1:
		stops = random.choice(destinations)
	else:
		stops = []

	aircraft = random.choice(aircrafts)
	return {
			"id": _id,
			"departure": departure,
			"arrival": arrival,
			"duration": duration,
			"airline": airline,

			"from": _from,
			"to": _to,
			"stops": stops,
			"status": "Confirmed",
			"aircraft": aircraft
	}

if __name__ == '__main__':
	print 'For testing purposes, have this generated entry'
	print '*' * 20
	print json.dumps(generate_entry()[1])

File: node/generate.js
var _ = require('underscore');
var genie = require('./genie');
var db = require('riak-js').getClient({
  pool: {
    servers: ['bdc-n-e3.fbi.h-da.de:8098', 'bdc-n-e4.fbi.h-da.de:8098']
  }
});
/*var db = require('riak-js').getClient({
  host: 'bdc-n-e3.fbi.h-da.de', port: '8098', api: 'protobuf'
});*/

var requests = 0;
var index = 0;

var MAX_REQUESTS = 100;

var instrument = {
  'riak.request.start': function(event) {
    console.log('[riak-js] [start] ' + event.method.toUpperCase() + ' ' + event.path);
    requests++;
  },
  'riak.request.end': function(event) {
    console.log('[riak-js] [end] ' + event.method.toUpperCase() + ' ' + event.path);
    requests--;
  }
}

db.registerListener(instrument);

var start = process.hrtime();

var elapsed_time = function(note) {
  var precision = 3; // 3 decimal places
  var elapsed = process.hrtime(start)[1] / 1000000; // divide by a million to get nano to milli
  console.log(process.hrtime(start)[0] + " s, " + elapsed.toFixed(precision) + " ms - " + note); // print message + time
  start = process.hrtime(); // reset the timer
}

var companies = ['Aeroflot', 'Air France', 'German Wings', 'Air Berlin', 'Lufthansa', 'Aer Lingus',
    'US Airways', 'Delta Airlines'
];
var destinations = [{
    "country": "France",
    "city": "Paris",
    "iata": "ORY",
    "airport": "Orly Arpt",
    "terminals": ["W"]
  }, {
    "country": "France",
    "city": "Montpellier",
    "iata": "MPL",
    "airport": "Frejorgues Arpt",
    "terminals": []
  }, {
    "country": "Germany",
    "city": "Munich",
    "iata": "MUC",
    "airport": "Munich Intl Arpt",
    "terminal": "2"
  }, {

    "country": "Germany",
    "city": "Dusseldorf",
    "iata": "DUS",
    "airport": "Dusseldorf Arpt",
    "terminal": ""
  }, {
    "country": "France",
    "city": "Toulouse",
    "iata": "TLS",
    "airport": "Blagnac Arpt",
    "terminal": ""
  }, {
    "country": "Thailand",
    "city": "Bangkok",
    "iata": "BKK",
    "airport": "Suvarnabhumi Intl Arpt",
    "terminal": ""
  }
];
var aircrafts = ['Airbus A380', 'Boeing 747', 'Airbus A320-200'];

var flightTemplate = {
  "id": {
    format: '##.#.#'
  },
  "personalFlightData": {
    max: 2,
    template: {
      eTicket: {
        format: '############C#'
      },
      "class": {
        oneOf: [
            'Economy',
            'First Class'
        ]
      },
      baggage: {
        format: 'ADT #PC'
      }
    }
  },
  "passengers": {
    max: 1,
    template: {
      "type": {
        format: '0'
      },
      "name": {
        pattern: 'firstName'
      },
      "surname": {
        pattern: 'lastName'
      },
      "id": function() {
        return this.name.toUpperCase() + '/' + this.surname.toUpperCase();
      }
    }
  },
  "flights": {
    min: 1,
    max: 4,
    template: {
      "id": {
        format: '..####',
      },
      "departure": function() {
        return Date.now() + 86400000 * _.random(0, 6);
      },
      "arrival": function() {
        return this.departure + 6 * _.random(30, 480);
      },
      "company": {
        oneOf: companies
      },
      "from": {
        oneOf: destinations
      },
      "to": {
        oneOf: destinations
      },
      "duration": function() {
        return this.arrival - this.departure;
      },
      "status": {
        oneOf: [
            'confirmed',
            'unconfirmed',
            'TERRORIST'
        ]
      },
      "aircraft": {
        oneOf: aircrafts
      }
    }
  },
  "fares": {
    template: {
      "type": function() {
        return 'ADT';
      },
      "fop": function() {
        return 'invoice'
      },
      "base": {
        range: [40, 200],
        places: 2
      },
      "taxes": {
        range: [10, 130],
        places: 2
      },
      total: function() {
        return this.base + this.taxes;
      },
      "currency": function() {
        return 'EUR';
      }
    }
  }
};

var args = process.argv.splice(2);
var amount = Number(args[0]) || 1;
var upload = Boolean(args[1]) || false;
var bucket = args[2];

if (args.length < 3) {
  console.log("USE: node generate [amount] [upload] [bucket] > [logfile]");
  console.log("     node generate 1000 true TestBucket > generate.log");
  process.exit();
}

console.log("Generating " + amount + " data sets...");

function generateData() {
  var flight = genie(flightTemplate);

  var segment = '1';
  for (var j = 1; j < flight.flights.length; j++) {
    var newSegment = j + 1;
    segment += ', ' + newSegment;
  };
  flight.fares.segment = segment;

  if (!upload) {
    console.dir(flight);
  } else {
    db.save(bucket, flight.id, flight);
  }

  if (index % 100 === 0) {
    console.log("Generated already " + index + " data sets...");
  }

  index++;
}

// checks that we don't overwhelm the server with requests
function checkRequests() {
  if (index < amount) {
    if (requests < MAX_REQUESTS) {
      generateData();
    } else {
      console.log("I have to wait for reqs to finish...");
    }
  } else {
    console.log("Finally finished...");
    process.exit();
  }
};

process.on('exit', function() {
  elapsed_time('Execution time for: ' + amount + ' data sets.');
});

process.on('SIGINT', function() {
  process.exit();
});

// start the whole mess
setInterval(checkRequests, 10);